File: events/2015/10/solution.py
import sys
import json
import time
from datetime import timedelta, datetime
from pathlib import Path

# Add workspace root to path for imports
workspace_root = Path(__file__).parent.parent.parent.parent
sys.path.insert(0, str(workspace_root))

from utils.python.types import AOCDayResults, AOCTimestamp, AOCPartResult
from utils.python.input import readInputAsArray
from utils.python.performance import measurePerformance

def countDigits(input: str) -> str:
  output = ""

  count = 0
  lastDigit = ""
  for digit in input:
    if count == 0:
      count = 1
      lastDigit = digit
      continue
    
    if digit == lastDigit:
      count += 1
    else:
      output += str(count) + lastDigit
      count = 1
      lastDigit = digit
  output += str(count) + lastDigit

  return output
# ----------------------------------------------------------------------------------------------------
# | Part 1
# ----------------------------------------------------------------------------------------------------
def part1(input: list[str]) -> int:
  lastIteration = input[0]

  for i in range(40):
    lastIteration = countDigits(lastIteration)

  return len(lastIteration)

# ----------------------------------------------------------------------------------------------------
# | Part 2
# ----------------------------------------------------------------------------------------------------
def part2(input: list[str]) -> int:
  lastIteration = input[0]

  for i in range(50):
    lastIteration = countDigits(lastIteration)

  return len(lastIteration)

# ----------------------------------------------------------------------------------------------------
# | Main Function
# ----------------------------------------------------------------------------------------------------
def main():
  # Get the arguments
  args = sys.argv[1:]
  fileName = args[0] if args else "input.txt"

  # Start the timer
  start_timestamp = datetime.now().isoformat()
  start_time = time.monotonic()

  # Read the input to an array of strings
  input = readInputAsArray(fileName)

  ## Run the parts
  p1Result = measurePerformance(part1, input)
  p2Result = measurePerformance(part2, input)

  # End the timer
  end_time = time.monotonic()
  end_timestamp = datetime.now().isoformat()
  duration = timedelta(seconds=end_time - start_time).total_seconds()*1000000000 # Convert to nanoseconds

  # Return the results
  return AOCDayResults(
    Year=2015,
    Day=10,
    Part1=p1Result,
    Part2=p2Result,
    Duration=duration,
    Timestamp=AOCTimestamp(
      Start=start_timestamp,
      End=end_timestamp,
    ),
  )



if __name__ == "__main__":
  results = main()
  print(json.dumps(results))

File: events/2015/10/solution.ts
import { AOCDayResults } from "@repo/utils-ts/returnType";
import { readInputAsArray } from "@repo/utils-ts/readInput";
import { measurePerformance } from "@repo/utils-ts/measurePerformance";

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
function countDigits(input: string): string {
  let output = ""

  let count = 0;
  let lastDigit = ""
  for (var i = 0; i < input.length; i++) {
    if (i == 0) {
      count = 1;
      lastDigit = input[0]
      continue;
    }

    let currentDigit = input[i];
    if (currentDigit == lastDigit) {
      count += 1
    } else {
      output += String(count) + lastDigit;
      count = 1;
      lastDigit = currentDigit;
    }
  }
  output += String(count) + lastDigit;

  return output;
}

function part1(input: string[]): number | undefined {
  let lastIteration = input[0];

  for (var i = 0; i < 40; i++) {
    lastIteration = countDigits(lastIteration)
  }

  return lastIteration.length
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
function part2(input: string[]): number | undefined {
  let lastIteration = input[0];

  for (var i = 0; i < 50; i++) {
    lastIteration = countDigits(lastIteration)
  }

  return lastIteration.length
}


// ----------------------------------------------------------------------------------------------------
// | Solve the puzzle
// ----------------------------------------------------------------------------------------------------
function solve() {
  // Get the arguments
  const args = process.argv.slice(2);
  const fileName = args[0] || "input.txt";

  // Start the timer
  const startTimeStamp = new Date(Date.now()).toISOString();
  const startTime = performance.now();

  // Read the input as an array of strings
  const input = readInputAsArray(fileName);

// Run the parts
const p1Result = measurePerformance(() => part1(input));
const p2Result = measurePerformance(() => part2(input));

// End the timer
const endTimeStamp = new Date(Date.now()).toISOString();
const endTime = performance.now();
const duration = endTime - startTime;

// Return the results
return {
  Year: 2015,
  Day: 10,
  Part1: p1Result,
  Part2: p2Result,
  Duration: Math.round(duration * 1000000), // Convert to nanoseconds
  Timestamp: {
    Start: new Date(startTimeStamp).toISOString(),
    End: new Date(endTimeStamp).toISOString(),
  },
} as AOCDayResults;
  
}
const results = solve();
console.log(JSON.stringify(results, null, 2));